File: src/dotnet/EntityFX.CrystalMarkNet/FPU/MikoFpu.cs
using System.Threading;

namespace EntityFX.CrystalMarkNet.FPU
{
    class MikoFpu : CrystalBenchmarkBase
    {
        private const int MemSize = 2048;

        protected override double BenchImplementation(CancellationToken cancellationToken)
        {
            double count = 0;

            float a = 1.0f;
            float b = 0.0f;
            float c = 0.0f;
            float d = 0.0f;

            float xch = 0.0f;

            for (; ; )
            {

                b += a;
                a -= d;

                xch = c;
                c = a;
                a = xch;

                a += d;

                xch = c;
                c = a;
                a = xch;

                a *= b;

                count++;

                if (cancellationToken.IsCancellationRequested)
                {
                    return count / 30000;
                }
            }
        }
    }
}

File: src/lua/fpu/mikofpu.lua
MikoFpu = class(CrystalBenchmarkBase, function(mikofpu, writer)
    CrystalBenchmarkBase.init(mikofpu, writer, "MikoFpu")
end)

MikoFpu.MemSize = 2048

function MikoFpu:benchImplementation()
    local count = 0

    local a = 1.0
    local b = 0.0
    local c = 0.0
    local d = 0.0

    local xch = 0.0

    while not self:isCancelled() do
        b = b + a
        a = a - d

        xch = c
        c = a
        a = xch

        a = a + d

        xch = c
        c = a
        a = xch

        a = a * b

        count = count + 1
        boost = count % 20
    end

    return count / 30000
end
